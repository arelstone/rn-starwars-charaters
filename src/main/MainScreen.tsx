import { StackNavigationProp } from '@react-navigation/stack';
import React, { FC } from 'react';
import { Text } from 'react-native';
import Screen from '../components/Screen/Screen';
import { AppNavigatorParams } from '../navgators';

interface MainScreenProps {
    navigation: StackNavigationProp<AppNavigatorParams, 'Main'>;
}

const MainScreen: FC<MainScreenProps> = () => {
    return <Screen>
        <Text>MainScreen</Text>
    </Screen>;
};

export default MainScreen;
