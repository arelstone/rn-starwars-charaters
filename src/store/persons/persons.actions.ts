import { Person, ResourceType } from '../../SwApi';
import { getNumber, nextPage } from '../../utils/storeUtils';
import { FAILED_FETCHING_PERSONS, START_FETCHING_PERSONS, SET_PERSONS, DONE_FETCHING_PERSONS } from './persons.types';
import { baseUrl } from '../../config';

export const fetchPersons = (page = 1, uri: ResourceType = ResourceType.People) => async (dispatch: any) => {
    if (page === 1) {
        dispatch({ type: START_FETCHING_PERSONS });
    }

    const response = await fetch(`${baseUrl()}/${uri}/?page=${page}`)
        .then(r => r.json())
        .catch(error => dispatch({
            type: FAILED_FETCHING_PERSONS,
            payload: error.message,
        }));

    const next = nextPage(response);

    dispatch({
        type: SET_PERSONS,
        payload: {
            nextPage: next,
            count: response.count,
            results: response.results.map((person: Person) => ({
                ...person,
                id: getNumber(person.url),
                homeworld_id: getNumber(person.homeworld),
            }))
                .sort((a: Person, b: Person) => a.name.toLowerCase().localeCompare(b.name.toLowerCase())),
        },
    });

    if (next) {
        return dispatch(fetchPersons(next));
    }
    console.log('###HERE', next);


    return dispatch({ type: DONE_FETCHING_PERSONS });
};
