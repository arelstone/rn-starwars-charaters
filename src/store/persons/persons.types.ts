import { Person } from '../../SwApi';

export const START_FETCHING_PERSONS = 'START_FETCHING_PERSONS';
export const DONE_FETCHING_PERSONS = 'DONE_FETCHING_PERSONS';
export const SET_PERSONS = 'SET_PERSONS';
export const FAILED_FETCHING_PERSONS = 'FAILED_FETCHING_PERSONS';

export type SetPersonsParams = {
    count: number;
    nextPage: number;
    results: Person[];
}

export type FetchPersonsAction = {
    type: PersonActionTypes;
}

type PersonActionTypes = typeof START_FETCHING_PERSONS
    | typeof DONE_FETCHING_PERSONS
    | typeof FAILED_FETCHING_PERSONS;

export type SetPersonsAction = {
    type: typeof SET_PERSONS;
    payload: SetPersonsParams;
}
