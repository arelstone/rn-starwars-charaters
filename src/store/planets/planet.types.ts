import { Planet } from '../../SwApi';

export const START_FETCHING_PLANETS = 'START_FETCHING_PLANETS';
export const DONE_FETCHING_PLANETS = 'DONE_FETCHING_PLANETS';
export const SET_PLANETS = 'SET_PLANETS';
export const FAILED_FETCHING_PLANETS = 'FAILED_FETCHING_PLANETS';

export type SetPlanetsParams = {
    count: number;
    nextPage: number;
    results: Planet[];
}

type PlanetActionTypes = typeof SET_PLANETS
    | typeof DONE_FETCHING_PLANETS
    | typeof START_FETCHING_PLANETS
    | typeof FAILED_FETCHING_PLANETS;

export type FetchPlanetsAction = {
    type: PlanetActionTypes;
}
