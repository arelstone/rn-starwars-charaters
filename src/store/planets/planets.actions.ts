import { Dispatch } from 'react';
import { GetState } from '..';
import { Planet } from '../../SwApi';
import { getNumber } from '../../utils/storeUtils';
import { SetPlanetsParams, FetchPlanetsAction } from './planet.types';

export const fetchPlanets = (page = 1) => async (dispatch: Dispatch<FetchPlanetsAction>, getState: GetState) => {
    if (page === 1) {
        dispatch({ type: 'START_FETCHING_PLANETS' });
    }
    const response = await fetch(`https://swapi.dev/api/planets/?page=${page}`)
        .then(r => r.json())
        .catch(error => console.log('###error', error));

    const nextPage = response.next && getNumber(response.next);

    const results = response.results.map((result: Planet) => ({
        ...result,
        id: getNumber(result.url),
    }));

    dispatch(setPlanets({
        nextPage: response.next && getNumber(response.next),
        results: [...results, ...getState().planets.results],
        count: response.count,
    }));

    // if (nextPage) {
    //     return dispatch(fetchPlanets(nextPage));
    // }

    return dispatch({ type: 'DONE_FETCHING_PLANETS' });
};

const setPlanets = (payload: SetPlanetsParams) => dispatch => {
    return dispatch({
        type: 'SET_PLANETS',
        payload,
    });
};
