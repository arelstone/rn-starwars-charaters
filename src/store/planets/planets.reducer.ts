import { Planet } from '../../SwApi';
import { START_FETCHING_PLANETS, DONE_FETCHING_PLANETS, SET_PLANETS, FAILED_FETCHING_PLANETS } from './planet.types';
import { StoreResource } from '../root.reducer';


export const INITIAL_STATE: StoreResource<Planet[]> = {
    count: 0,
    loading: false,
    nextPage: undefined,
    results: [],
    error: {
        hasError: false,
        message: '',
    },
};

export default (state = INITIAL_STATE, { type, payload }: any): StoreResource<Planet[]> => {
    switch (type) {
        case START_FETCHING_PLANETS:
            return {
                ...state,
                results: [],
                loading: true,
                error: {
                    hasError: false,
                    message: '',
                },
            };
        case DONE_FETCHING_PLANETS:
            return {
                ...state,
                loading: false,
            };
        case FAILED_FETCHING_PLANETS:
            return {
                ...state,
                loading: false,
                error: {
                    hasError: true,
                    message: payload,
                },
            };
        case SET_PLANETS: {
            return {
                ...state,
                ...payload,
                results: [
                    ...state.results,
                    ...payload.results,
                ],
            };
        }
        default:
            return state;
    }
};

