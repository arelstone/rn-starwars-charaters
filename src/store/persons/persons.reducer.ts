import { Person } from '../../SwApi';
import { StoreResource } from '../root.reducer';
import { DONE_FETCHING_PERSONS, FAILED_FETCHING_PERSONS, SET_PERSONS, START_FETCHING_PERSONS } from './persons.types';

export const INITIAL_STATE: StoreResource<Person[]> = {
    count: 0,
    loading: false,
    error: {
        hasError: false,
        message: '',
    },
    nextPage: undefined,
    results: [],
};

export default (state = INITIAL_STATE, { type, payload }: any): StoreResource<Person[]> => {
    switch (type) {
        case START_FETCHING_PERSONS:
            return {
                ...state,
                results: [],
                loading: true,
                error: {
                    hasError: false,
                    message: '',
                },
            };
        case DONE_FETCHING_PERSONS:
            return {
                ...state,
                loading: false,
            };
        case FAILED_FETCHING_PERSONS:
            return {
                ...state,
                loading: false,
                error: {
                    hasError: true,
                    message: payload,
                },
            };
        case SET_PERSONS: {
            return {
                ...state,
                ...payload,
                results: [
                    ...state.results,
                    ...payload.results,
                ],
            };
        }
        default:
            return state;
    }
};

