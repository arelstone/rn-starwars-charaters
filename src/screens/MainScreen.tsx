import React, { FC, useEffect, useRef, useState } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import { ListRenderItemInfo, Text, View, FlatList } from 'react-native';
import { connect, DispatchProp } from 'react-redux';
import { AppNavigatorParams } from '../navgators';
import { RootState } from '../store';
import { fetchPlanets } from '../store/planets/planets.actions';
import { fetchPersons } from '../store/persons/persons.actions';
import { Person } from '../SwApi';
import { Screen, Loading } from '../components';
import PersonCard from '../components/PersonCard';

interface MainScreenProps extends DispatchProp<any> {
    navigation: StackNavigationProp<AppNavigatorParams, 'Main'>;
    charaters: RootState['persons'];
    planets: RootState['planets'];
}

export const NUMBER_OF_COLUMNS = 3;

const MainScreen: FC<MainScreenProps> = ({ navigation: { navigate }, charaters, dispatch, planets }) => {
    const [isLoading] = useState<boolean>(planets.loading || charaters.loading);

    useRef(useEffect(() => {
        dispatch(fetchPersons());
        dispatch(fetchPlanets());
    }, []));

    console.log(planets.results[0]);

    const hasError = planets.error.hasError || charaters.error.hasError;
    const errorMessage = planets.error.message || charaters.error.message;

    if (isLoading) {
        return <Loading />;
    }

    return <Screen>

        {hasError && <View><Text>{errorMessage}</Text></View>}

        <FlatList
            contentContainerStyle={{ padding: 10 }}
            data={charaters.results || []}
            renderItem={({ item }: ListRenderItemInfo<Person>) => <PersonCard
                charater={item}
                onPress={() => navigate('Details', { id: item.id })}
            />}
            numColumns={NUMBER_OF_COLUMNS}
            keyExtractor={item => `item_${item.id}`}
        />
    </Screen>;
};

export default connect(
    (state: RootState) => ({
        charaters: state.persons,
        planets: state.planets,
    }),
)(MainScreen);
