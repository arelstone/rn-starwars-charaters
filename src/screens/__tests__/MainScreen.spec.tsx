import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { MainScreen } from '../MainScreen';
import { personsMock } from '../../../__mocks__/persons.mock';
import { planetsMock } from '../../../__mocks__/planets.mock';
import { navigationMock } from '../../../__mocks__/navigation.mock';

let wrapper: ShallowWrapper;
const mockDispatch = jest.fn();

const charaters = {
    count: 2,
    error: {
        hasError: false, message: '',
    },
    loading: false,
    nextPage: undefined,
    results: personsMock,
};

const planets = {
    count: 2,
    error: { hasError: false, message: '' },
    loading: false,
    nextPage: undefined,
    results: planetsMock,
};

describe('<MainScreen />', () => {
    beforeAll(() => {
        wrapper = shallow(<MainScreen
            charaters={charaters}
            planets={planets}
            dispatch={mockDispatch}
            navigation={navigationMock}
        />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe('FlatList', () => {
        it('should contain a FlatList', () => {
            expect(wrapper.find('FlatList').exists()).toBeTruthy();
        });

        it('should have correct data', () => {
            expect(wrapper.find('FlatList').prop('data')).toEqual(personsMock);
        });
    });

    describe('Loading', () => {
        it('should render Loading when CHARATERS is loading', () => {
            wrapper.setProps({
                charaters: { ...charaters, loading: true },
            });

            expect(wrapper.find('Loading').exists()).toBeTruthy();
        });

        it('should render Loading when PLANETS is loading', () => {
            wrapper.setProps({
                charaters: { ...charaters, loading: false },
                planets: { ...planets, loading: true },
            });

            expect(wrapper.find('Loading').exists()).toBeTruthy();
        });

        it('should render Loading when both PLANETS and CHARATERS is loading', () => {
            wrapper.setProps({
                charaters: { ...charaters, loading: true },
                planets: { ...planets, loading: true },
            });

            expect(wrapper.find('Loading').exists()).toBeTruthy();
        });

        it('should NOT render loading when both PLANETS and CHARATERS is NOT loading', () => {
            wrapper.setProps({
                charaters: { ...charaters, loading: false },
                planets: { ...planets, loading: false },
            });

            expect(wrapper.find('Loading').exists()).toBeFalsy();
        });
    });

    describe('onPress', () => {
        it('should invoke onPress', () => {
            const item = personsMock[0];

            // eslint-disable-next-line
            // @ts-expect-error
            wrapper.find('FlatList').prop('renderItem')({ item }).props.onPress();

            expect(navigationMock.navigate).toHaveBeenCalledWith('Details', { id: item.id, planetId: item.homeworld_id });
        });
    });

});
