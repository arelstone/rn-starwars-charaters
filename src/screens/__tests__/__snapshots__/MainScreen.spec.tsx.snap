// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainScreen /> should match snapshot 1`] = `
<Screen>
  <FlatList
    contentContainerStyle={
      Object {
        "padding": 10,
      }
    }
    data={
      Array [
        Object {
          "birth_year": "19 BBY",
          "created": "2014-12-09T13:50:51.644000Z",
          "edited": "2014-12-10T13:52:43.172000Z",
          "eye_color": "Blue",
          "films": Array [],
          "gender": "male",
          "hair_color": "Blond",
          "height": "172",
          "homeworld": "https://swapi.dev/api/planets/1/",
          "homeworld_id": 1,
          "id": 1,
          "mass": "77",
          "name": "Luke Skywalker",
          "skin_color": "Fair",
          "species": Array [],
          "starships": Array [],
          "url": "https://swapi.dev/api/people/1/",
          "vehicles": Array [
            "https://swapi.dev/api/vehicles/14/",
          ],
        },
        Object {
          "birth_year": "112BBY",
          "created": "2014-12-10T15:10:51.357000Z",
          "edited": "2014-12-20T21:17:50.309000Z",
          "eye_color": "yellow",
          "films": Array [],
          "gender": "n/a",
          "hair_color": "n/a",
          "height": "167",
          "homeworld": "http://swapi.dev/api/planets/1/",
          "homeworld_id": 1,
          "id": 2,
          "mass": "75",
          "name": "C-3PO",
          "skin_color": "gold",
          "species": Array [],
          "starships": Array [],
          "url": "http://swapi.dev/api/people/2/",
          "vehicles": Array [],
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Screen>
`;
